### PR TITLE
New labels format in json 

### PR DESCRIPTION
**Pull request by SalvaMilenario**
|Related issue|
|---|
|#3429|

## Description
The format of the labels is changed so that they appear by levels as it is encoded in JSON.

The new format shows in addition to the name of the key and its value tagged as such, it has been added to show if it is a hidden label.

Additionally, the expected behavior when saving two labels with the same key is that only the second one that is entered is saved.

The format with which the labels are displayed is:

``` JSON
# curl -u foo:bar -k -X GET "http://172.16.1.13:55000/agents/001/config/agent/labels?pretty"
{
   "error": 0,
   "data": {
      "labels": [
         {
            "value": "i-001c",
            "key": "id",
            "hidden": "no"
         },
         {
            "value": "default",
            "key": "group",
            "hidden": "no"
         },
         {
            "value": "172.16.1.14",
            "key": "ip",
            "hidden": "no"
         },
         {
            "value": "Sep 18th, 2019",
            "key": "installation",
            "hidden": "yes"
         }
      ]
   }
}
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows 
- Memory tests
  - [x] Valgrind report for affected components
  - [x] Scan-build
  - [x] Address Sanitizer

### CODE REVIEW

File: src/client-agent/config.c [hunk 1]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
 
     unsigned int i;
     cJSON *root = cJSON_CreateObject();
-    cJSON *labels = cJSON_CreateObject();
+    cJSON *label_s = cJSON_CreateArray();
+    cJSON *label = cJSON_CreateObject();
+    cJSON *labels = cJSON_CreateArray();
 
     if (agt->labels) {
         for (i=0;agt->labels[i].key;i++) {
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Reduce the scope of `i` variable.

File: src/client-agent/config.c [hunk 2]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
...
-            cJSON_AddStringToObject(labels,agt->labels[i].key,agt->labels[i].value);
+            cJSON *label_k = cJSON_CreateObject();
+            cJSON_AddStringToObject(label_k,"value", agt->labels[i].value);
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Add spaces between parameters as in `param1, param2, ..., paramN`. Applies to the whole PR.

File: src/client-agent/config.c [hunk 3]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
...
+            cJSON_AddStringToObject(label_k,"key", agt->labels[i].key);
+            if (agt->labels[i].flags.hidden)
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Add bracers to every block.

File: src/client-agent/config.c [hunk 4]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
 
     unsigned int i;
     cJSON *root = cJSON_CreateObject();
-    cJSON *labels = cJSON_CreateObject();
+    cJSON *label_s = cJSON_CreateArray();
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Use meaningful variable names. And reduce variable scope as much as possible.

File: src/client-agent/config.c [hunk 5]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
...
-            cJSON_AddStringToObject(labels,agt->labels[i].key,agt->labels[i].value);
+            cJSON *label_k = cJSON_CreateObject();
+            cJSON_AddStringToObject(label_k,"value", agt->labels[i].value);
+            cJSON_AddStringToObject(label_k,"key", agt->labels[i].key);
+            if (agt->labels[i].flags.hidden)
+                cJSON_AddStringToObject(label_k,"hidden", "yes");
+            if (agt->labels[i].flags.system)
+                cJSON_AddStringToObject(label_k,"system", "yes");
+            cJSON_AddItemToObject(label_s,"",label_k);
+            free(label_k);
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Objects added to another object using `cJSON_AddItemToObject` must not be freed, as JSON_Delete will take care of freeing them. 
That `free` breaks the parent object.

File: src/client-agent/config.c [hunk 6]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
...
+    cJSON *labels = cJSON_CreateArray();
 
     if (agt->labels) {
         for (i=0;agt->labels[i].key;i++) {
-            cJSON_AddStringToObject(labels,agt->labels[i].key,agt->labels[i].value);
+            cJSON *label_k = cJSON_CreateObject();
+            cJSON_AddStringToObject(label_k,"value", agt->labels[i].value);
+            cJSON_AddStringToObject(label_k,"key", agt->labels[i].key);
+            if (agt->labels[i].flags.hidden)
+                cJSON_AddStringToObject(label_k,"hidden", "yes");
```
Comment (JuantAldea) [inline src/client-agent/config.c]: This lacking the `hidden: no` option?

File: src/client-agent/config.c [hunk 7]
```diff
@@ -128,14 +128,26 @@ cJSON *getLabelsConfig(void) {
...
+            if (agt->labels[i].flags.system)
```
Comment (JuantAldea) [inline src/client-agent/config.c]: As per a conversation with @chemamartinez, system flags should not appear.
Configuration is being read here, and system flags are not part of the configuration. 
See https://documentation.wazuh.com/3.9/user-manual/reference/ossec-conf/labels.html

File: src/client-agent/config.c [hunk 8]
```diff
@@ -126,17 +126,31 @@ cJSON *getLabelsConfig(void) {
...
     if (agt->labels) {
-        for (i=0;agt->labels[i].key;i++) {
-            cJSON_AddStringToObject(labels,agt->labels[i].key,agt->labels[i].value);
+        unsigned int i;
+        for (i=0; agt->labels[i].key; i++) {
+            cJSON *label = cJSON_CreateObject();
+            cJSON_AddStringToObject(label, "value", agt->labels[i].value);
+            cJSON_AddStringToObject(label, "key", agt->labels[i].key);
+
+            if (agt->labels[i].flags.hidden) {
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Change to `cJSON_AddStringToObject(label, "hidden", agt->labels[i].flags.hidden ? "yes" : "no");`

File: src/client-agent/config.c [hunk 9]
```diff
@@ -126,17 +126,31 @@ cJSON *getLabelsConfig(void) {
         return NULL;
     }
 
-    unsigned int i;
     cJSON *root = cJSON_CreateObject();
-    cJSON *labels = cJSON_CreateObject();
+    cJSON *labels = cJSON_CreateArray();
```
Comment (JuantAldea) [inline src/client-agent/config.c]: Reduce variable scope of these variables as much as possible.

**Comment by SalvaMilenario** [comment]
Hi team

This is logs of the valgrind:
```log
==9035== 
==9035== HEAP SUMMARY:
==9035==     in use at exit: 93,264 bytes in 57 blocks
==9035==   total heap usage: 6,308 allocs, 6,251 frees, 3,260,632 bytes allocated
==9035== 
==9035== 272 bytes in 1 blocks are possibly lost in loss record 37 of 48
==9035==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==9035==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==9035==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==9035==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==9035==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==9035==    by 0x465B90: CreateThreadJoinable (pthreads_op.c:47)
==9035==    by 0x465C27: CreateThread (pthreads_op.c:62)
==9035==    by 0x406606: AgentdStart (agentd.c:127)
==9035==    by 0x40909F: main (main.c:179)
==9035== 
==9035== 272 bytes in 1 blocks are possibly lost in loss record 38 of 48
==9035==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==9035==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==9035==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==9035==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==9035==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==9035==    by 0x465B90: CreateThreadJoinable (pthreads_op.c:47)
==9035==    by 0x465C27: CreateThread (pthreads_op.c:62)
==9035==    by 0x40664F: AgentdStart (agentd.c:136)
==9035==    by 0x40909F: main (main.c:179)
==9035== 
==9035== 272 bytes in 1 blocks are possibly lost in loss record 39 of 48
==9035==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==9035==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==9035==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==9035==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==9035==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==9035==    by 0x465B90: CreateThreadJoinable (pthreads_op.c:47)
==9035==    by 0x465C27: CreateThread (pthreads_op.c:62)
==9035==    by 0x40677B: AgentdStart (agentd.c:150)
==9035==    by 0x40909F: main (main.c:179)
==9035== 
==9035== 272 bytes in 1 blocks are possibly lost in loss record 40 of 48
==9035==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==9035==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==9035==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==9035==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==9035==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==9035==    by 0x465B90: CreateThreadJoinable (pthreads_op.c:47)
==9035==    by 0x465C27: CreateThread (pthreads_op.c:62)
==9035==    by 0x4068C8: AgentdStart (agentd.c:188)
==9035==    by 0x40909F: main (main.c:179)
==9035== 
==9035== LEAK SUMMARY:
==9035==    definitely lost: 0 bytes in 0 blocks
==9035==    indirectly lost: 0 bytes in 0 blocks
==9035==      possibly lost: 1,088 bytes in 4 blocks
==9035==    still reachable: 92,176 bytes in 53 blocks
==9035==         suppressed: 0 bytes in 0 blocks
==9035== Reachable blocks (those to which a pointer was found) are not shown.
==9035== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==9035== 
==9035== For counts of detected and suppressed errors, rerun with: -v
==9035== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
```
The compilation supervised with `scan-build` not notificated error in the changed files. In addition, `address sanitizer` not report others memory leak.

**Comment by SalvaMilenario** [comment]
The JSON that is generated in the function `getLabelsConfig` it's the same that the API shows with it.

```
# curl -u foo:bar -k -X GET "http://172.16.1.13:55000/agents/002/config/agent/labels?pretty"
{
   "error": 0,
   "data": {
      "labels": [
         {
            "label": [
               {
                  "value": "002",
                  "key": "id",
                  "hidden": "no"
               },
               {
                  "value": "default",
                  "key": "group",
                  "hidden": "no"
               },
               {
                  "value": "172.16.1.18",
                  "key": "ip",
                  "hidden": "no"
               },
               {
                  "value": "Sep. 16th, 2019",
                  "key": "installation",
                  "hidden": "yes"
               }
            ]
         }
      ]
   }
}
```
```
2019/09/16 10:12:58 ossec-agentd: INFO: {
	"labels":	[{
			"label":	[{
					"value":	"002",
					"key":	"id",
					"hidden":	"no"
				}, {
					"value":	"default",
					"key":	"group",
					"hidden":	"no"
				}, {
					"value":	"172.16.1.18",
					"key":	"ip",
					"hidden":	"no"
				}, {
					"value":	"Sep. 16th, 2019",
					"key":	"installation",
					"hidden":	"yes"
				}]
		}]
}

```

**Comment by chemamartinez** [comment]
After discussing it with the @wazuh/framework, it doesn't make sense to include the `label` array since it will be always a unique structure for all the labels of an agent. The correct JSON output should look like:

```
# curl -u foo:bar -k -X GET "http://172.16.1.13:55000/agents/002/config/agent/labels?pretty"
{
   "error": 0,
   "data": {
      "labels": [
               {
                  "value": "002",
                  "key": "id",
                  "hidden": "no"
               },
               {
                  "value": "default",
                  "key": "group",
                  "hidden": "no"
               },
               {
                  "value": "172.16.1.18",
                  "key": "ip",
                  "hidden": "no"
               },
               {
                  "value": "Sep. 16th, 2019",
                  "key": "installation",
                  "hidden": "yes"
               }
      ]
   }
}
```

Please @SalvaMilenario apply the changes and test it works fine. Ping @wazuh/frontend as well to check it is a valid format for you to parse this JSON format to XML.

**Comment by SalvaMilenario** [comment]
Hi

I've changed the code with the new JSON format. 

The compilation supervised with `scan-build` doesn't notification any bugs, and the execution with `Valgrind` doesn't indicate any memory leaks. Besides, `address sanitizer` not report other memory leaks.

```
==5978== 
==5978== HEAP SUMMARY:
==5978==     in use at exit: 93,213 bytes in 55 blocks
==5978==   total heap usage: 41,430 allocs, 41,375 frees, 1,096,927,016 bytes allocated
==5978== 
==5978== 272 bytes in 1 blocks are possibly lost in loss record 37 of 48
==5978==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5978==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==5978==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==5978==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==5978==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==5978==    by 0x4690B4: CreateThreadJoinable (pthreads_op.c:47)
==5978==    by 0x46914B: CreateThread (pthreads_op.c:62)
==5978==    by 0x406620: AgentdStart (agentd.c:131)
==5978==    by 0x40910D: main (main.c:179)
==5978== 
==5978== 272 bytes in 1 blocks are possibly lost in loss record 38 of 48
==5978==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5978==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==5978==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==5978==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==5978==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==5978==    by 0x4690B4: CreateThreadJoinable (pthreads_op.c:47)
==5978==    by 0x46914B: CreateThread (pthreads_op.c:62)
==5978==    by 0x406669: AgentdStart (agentd.c:139)
==5978==    by 0x40910D: main (main.c:179)
==5978== 
==5978== 272 bytes in 1 blocks are possibly lost in loss record 39 of 48
==5978==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5978==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==5978==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==5978==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==5978==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==5978==    by 0x4690B4: CreateThreadJoinable (pthreads_op.c:47)
==5978==    by 0x46914B: CreateThread (pthreads_op.c:62)
==5978==    by 0x406795: AgentdStart (agentd.c:151)
==5978==    by 0x40910D: main (main.c:179)
==5978== 
==5978== 272 bytes in 1 blocks are possibly lost in loss record 40 of 48
==5978==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5978==    by 0x4012EE4: allocate_dtv (dl-tls.c:296)
==5978==    by 0x4012EE4: _dl_allocate_tls (dl-tls.c:460)
==5978==    by 0x57E1D92: allocate_stack (allocatestack.c:589)
==5978==    by 0x57E1D92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==5978==    by 0x4690B4: CreateThreadJoinable (pthreads_op.c:47)
==5978==    by 0x46914B: CreateThread (pthreads_op.c:62)
==5978==    by 0x4068F6: AgentdStart (agentd.c:189)
==5978==    by 0x40910D: main (main.c:179)
==5978== 
==5978== LEAK SUMMARY:
==5978==    definitely lost: 0 bytes in 0 blocks
==5978==    indirectly lost: 0 bytes in 0 blocks
==5978==      possibly lost: 1,088 bytes in 4 blocks
==5978==    still reachable: 92,125 bytes in 51 blocks
==5978==         suppressed: 0 bytes in 0 blocks
==5978== Reachable blocks (those to which a pointer was found) are not shown.
==5978== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==5978== 
==5978== For counts of detected and suppressed errors, rerun with: -v
==5978== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
```

Regards

**Comment by JcabreraC** [comment]
## Test

The API call works correctly and shows the new `labels` configuration format:

```JSON
# curl -u foo:bar -k -X GET "http://10.0.0.1:55000/agents/014/config/agent/labels?pretty"
{
   "error": 0,
   "data": {
      "labels": [
         {
            "value": "i-052a1838d",
            "key": "aws.instance-id",
            "hidden": "no"
         },
         {
            "value": "sg-1103",
            "key": "aws.sec-group",
            "hidden": "no"
         },
         {
            "value": "172.17.0.0",
            "key": "network.ip",
            "hidden": "no"
         },
         {
            "value": "02:42:ac:11:00:02",
            "key": "network.mac",
            "hidden": "no"
         },
         {
            "value": "January 1st, 2017",
            "key": "installation",
            "hidden": "yes"
         }
      ]
   }
}
```


This information is also displayed in the kibana app:

kibana labels

**Comment by chemamartinez** [comment]
Blocked by the @wazuh/frontend. The labels configuration should be displayed as follows for the new format:

image

**Comment by chemamartinez** [comment]
Blocked by the @wazuh/frontend. The labels configuration should be displayed as follows for the new format:

image

**Comment by juankaromo** [comment]
Working on it here https://github.com/wazuh/wazuh-kibana-app/issues/1816

**Comment by JcabreraC** [comment]
## Test
New JSON format added to the manager. The API call works correctly and shows the new `labels` configuration format:

```JSON
# curl -u foo:bar -k -X GET "http://10.0.0.1:55000/manager/config/analysis/labels?pretty"
{
   "error": 0,
   "data": {
      "labels": [
         {
            "value": "i-052a1838c",
            "key": "aws.instance-id",
            "hidden": "no"
         },
         {
            "value": "sg-1103",
            "key": "aws.sec-group",
            "hidden": "no"
         },
         {
            "value": "172.17.0.0",
            "key": "network.ip",
            "hidden": "no"
         },
         {
            "value": "02:42:ac:11:00:02",
            "key": "network.mac",
            "hidden": "no"
         },
         {
            "value": "January 1st, 2017",
            "key": "installation",
            "hidden": "yes"
         }
      ]
   }
}
```

**Pull request merged**